### PR TITLE
Adding set_config_entry

### PR DESCRIPTION
**Pull request by hkaiser**
- made key/value registry thread safe
- renamed the header file declaring the API functions
- add test

### CODE REVIEW

File: hpx/util/ini.hpp [hunk 1]
```diff
@@ -93,43 +124,117 @@ namespace hpx { namespace util
         void dump(int ind = 0, std::ostream& strm = std::cout) const;
 
         void add_section(std::string const& sec_name, section& sec,
-            section* root = nullptr);
-        bool has_section(std::string const& sec_name) const;
+            section* root = nullptr)
+        {
+            util::ignore_mutex_while_checking<mutex_type> im(mtx_);
```
Comment (sithhell) [inline hpx/util/ini.hpp]: Why do we want to ignore that lock?

Comment (hkaiser) [inline hpx/util/ini.hpp]: Hmmm, now as I think of it, we might get away without ignoring it. I'll play with this for a bit...

**Comment by hkaiser** [comment]
> Do we really initialize the config with multiple threads? I was under the impression that this would only always happen at startup and is then never changed.

This patch introduces a new API function `set_config_entry`. That means it can be called at any point in time.

**Comment by sithhell** [comment]
What are the use cases for `set_config_entry`? What are the semantics in terms of making the changes visible to the application, especially when multiple localities are involved?

**Comment by hkaiser** [comment]
> What are the use cases for `set_config_entry`? 

The use case is runtime adaptive tuning of parcel coalescing. We want to be able to control the parameters for this dynamically. 

> What are the semantics in terms of making the changes visible to the application, especially when multiple localities are involved?

Same as before: the changes are propagated only to the locality where `set_config_entry` is called. No news here.

**Comment by sithhell** [comment]
On 10/24/2016 02:31 PM, Hartmut Kaiser wrote:
>     Do we really initialize the config with multiple threads? I was
>     under the impression that this would only always happen at startup
>     and is then never changed.
> 
> This patch introduces a new API function |set_config_entry|. That means
> it can be called at any point in time.

Is this really what we want? What are the semantics here? What about
keeping the config consistent over localities? What's the usecase?

**Comment by sithhell** [comment]
On 10/25/2016 12:57 PM, Hartmut Kaiser wrote:
>     What are the use cases for |set_config_entry|?
> 
> The use case is runtime adaptive tuning of parcel coalescing. We want to
> be able to control the parameters for this dynamically.

Do you really want to go through the runtime system config there?
Wouldn't some plugin specific mechanism be more suitable? In the end,
setting/getting config entries is always going through a string, or
other form of type-erasure, which isn't exactly the fastest form to
store/retrieve values.

> 
>     What are the semantics in terms of making the changes visible to the
>     application, especially when multiple localities are involved?
> 
> Same as before: the changes are propagated only to the locality where
> |set_config_entry| is called. No news here.

With the difference that it is now exposed as part of the public API, so
the config is, potentially, subject to change after runtime system
startup. Especially for runtime adaptivity use case, having special
purpose mechanisms sounds like the better option instead.

**Comment by hkaiser** [comment]
> Do you really want to go through the runtime system config there?

Yes. a) it's already in place, b) why introduce something specific just for this? 

> In the end, setting/getting config entries is always going through a string, or other form of type-erasure, which isn't exactly the fastest form to store/retrieve values.

So what? Even if we have to go through a conversions to/from a string, if this is done once in a while only (perhaps once every 10 milliseconds) the introduced overhead shouldn't even be measurable.

> With the difference that it is now exposed as part of the public API, so the config is, potentially, subject to change after runtime system startup. 

Correct. Again: so what? I'd rather think this is a good thing (tm).

> Especially for runtime adaptivity use case, having special purpose mechanisms sounds like the better option instead.

So you're proposing to introduce a custom solution for each and every runtime adaptive use case we might come across? I don't think this is a scalable solution.

Another point speaking against a custom solution would be that the coalescing module is a plugin module, i.e. none of the binaries directly link to it. Introducing a custom solution for setting the parameters would require either a custom solution for loading this plugin or direct linking with it. Both options are not exactly what we want.

**Comment by sithhell** [comment]
Am 26.10.2016 6:56 nachm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>>
>> Do you really want to go through the runtime system config there?
>
> Yes. a) it's already in place, b) why introduce something specific just
for this?

Because every component will require a specific, well defined interface (in
one way or another). Why not stick with c++ and benefit from compile time
checks and native types?

>>
>> In the end, setting/getting config entries is always going through a
string, or other form of type-erasure, which isn't exactly the fastest form
to store/retrieve values.
>
> So what? Even if we have to go through a conversions to/from a string, if
this is done once in a while only (perhaps once every 10 milliseconds) the
introduced overhead shouldn't even be measurable.
>>
>> With the difference that it is now exposed as part of the public API, so
the config is, potentially, subject to change after runtime system startup.
>
> Correct. Again: so what? I'd rather think this is a good thing (tm).

For the record: I think such an API might indeed have it's benefits. For
this specific use case, it feels like a quick and dirty hack.

>>
>> Especially for runtime adaptivity use case, having special purpose
mechanisms sounds like the better option instead.
>
> So you're proposing to introduce a custom solution for each and every
runtime adaptive use case we might come across? I don't think this is a
scalable solution.

Yes, that's exactly what I'm proposing. Every component will need to have
it's interface defined anyway. I'd prefer a properly designed c++ mechanism
over runtime based string lookups.

>
> Another point speaking against a custom solution would be that the
coalescing module is a plugin module, i.e. none of the binaries directly
link to it. Introducing a custom solution for setting the parameters would
require either a custom solution for loading this plugin or direct linking
with it. Both options are not exactly what we want.

That's another discussion, I guess...

**Comment by hkaiser** [comment]
> That's another discussion, I guess...

It is not. This is exactly why we can't have a special interface just to set the values

**Comment by sithhell** [comment]
> > That's another discussion, I guess...
> 
> It is not. This is exactly why we can't have a special interface just to set
> the values

It is to the extent that whether the runtime config is the right place to 
exchange information for a feedback loop or not. We just don't know all 
requirements yet. It is also to the extent whether the parcel coalescing 
component should be a dynamically loadable module or not. It is also to the 
extent why the auto tuning feature of a specific component shouldn't be tightly 
coupled with the component it tries to adapt.

As said, having the an easy to use API function to set config values might 
indeed be helpful. I don't think it is the right tool for the auto tuning 
feature you described (not saying that it wouldn't work that way).

**Comment by sithhell** [comment]
On Mittwoch, 26. Oktober 2016 12:50:03 CEST Hartmut Kaiser wrote:
> > That's another discussion, I guess...
> 
> It is not. This is exactly why we can't have a special interface just to set
> the values

It is to the extent that whether the runtime config is the right place to 
exchange information for a feedback loop or not. We just don't know all 
requirements yet. It is also to the extent whether the parcel coalescing 
component should be a dynamically loadable module or not. It is also to the 
extent why the auto tuning feature of a specific component shouldn't be tightly 
coupled with the component it tries to adapt.

As said, having the an easy to use API function to set config values might 
indeed be helpful. I don't think it is the right tool for the auto tuning 
feature you described (not saying that it wouldn't work that way).

File: hpx/util/register_locks.hpp [hunk 2]
```diff
@@ -78,6 +78,23 @@ namespace hpx { namespace util
         void const* mtx_;
     };
 
+    template <typename Mutex>
+    struct ignore_mutex_while_checking
```
Comment (sithhell) [inline hpx/util/register_locks.hpp]: Is this still needed? After all, a mutex is something lockable, and this isn't used now anymore.

Comment (hkaiser) [inline hpx/util/register_locks.hpp]: Done.

**Pull request merged**